controller/backing_image_controller.go

```diff
@@ -291,9 +304,133 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
 		return err
 	}
 
+	if err := bic.replenishBackingImageCopies(backingImage); err != nil {
+		return err
+	}
+
+	bic.cleanupEvictionRequestedBackingImageCopies(backingImage)
+
+	return nil
+}
+
+func (bic *BackingImageController) replenishBackingImageCopies(bi *longhorn.BackingImage) (err error) {
+	bids, err := bic.ds.GetBackingImageDataSource(bi.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to get the backing image data source")
+	}
+	// only maintain the minNumberOfReplicas after BackingImage is transferred to BackingImageManager
+	if !bids.Spec.FileTransferred {
+		return nil
+	}
+
+	concurrentReplenishLimit, err := bic.ds.GetSettingAsInt(types.SettingNameConcurrentBackingImageReplenishPerNodeLimit)
+	if err != nil {
+		return err
+	}
+
+	// If disabled backing image replenishing, skip all the replenish except first copy.
+	if concurrentReplenishLimit == 0 {
+		return nil
+	}
+
+	nonFailedCopies := 0
+	for diskUUID := range bi.Spec.DiskFileSpecMap {
+		fileStatus, exists := bi.Status.DiskFileStatusMap[diskUUID]
+		if !exists || (fileStatus.State != longhorn.BackingImageStateFailed &&
+			fileStatus.State != longhorn.BackingImageStateFailedAndCleanUp &&
+			fileStatus.State != longhorn.BackingImageStateUnknown) {
+			// Non-existing file in status could due to not being synced from backing image manager yet.
+			// Consider it as newly created copy and count it as non-failed copies.
+			// So we don't create extra copy when handling copies evictions.
+			nonFailedCopies += 1
+		}
+	}
+
+	if nonFailedCopies == 0 {
+		return nil
+	} else if nonFailedCopies >= bi.Spec.MinNumberOfCopies {
+		if err := bic.handleBackingImageCopiesEvictions(nonFailedCopies, bi); err != nil {
+			return nil
+		}
```

Comment:
Do you mean
we count SatisfiedCopies here and then we can remove the logic of `handleBackingImageCopiesEvictions`?

Reply:
fixed.

Reply:
Wait this is not correct.
Because, if we count the SatisfiedCopies
- Non-failed &&
- Not eviction requested &&
- All Selector satisfied
Then we cannot distinguish the situation of SatisfiedCopies == 0
It can be 
1. there is no healthy copy, all the copies are dead, we should re-prepare the copy (for download and export-from-volume)
2. there is one copy, but it get eviction request.
    - In this case, we should first replenish one as backup before really delete the evicted copy

Thus, we can't mixed it together

Reply:
> there is one copy, but it get eviction request.

Nice catch. This is what I haven't considered.

NIT: And the next question is, does the controller need to count `if nonFailedCopies >= bi.Spec.MinNumberOfCopies`? I think it can directly do replenishment when `if NonEvictingCount < bi.Spec.MinNumberOfCopies`. I mean:
```
	if nonFailedCopies == 0 {
		return nil
	} else {
		// Get NonEvictingCount
		......
		if NonEvictingCount < bi.Spec.MinNumberOfCopies {
			log.Infof("Creating one more backing image copy for eviction")
			readyNode, readyDiskName, err := bic.ds.GetReadyNodeDiskForBackingImage(bi)
			if err != nil {
				logrus.WithError(err).Warnf("failed to create the backing image copy")
				return nil
			}
			// BackingImageManager will then sync the BackingImage to the disk
			bi.Spec.DiskFileSpecMap[readyNode.Status.DiskStatus[readyDiskName].DiskUUID] = &longhorn.BackingImageDiskFileSpec{}
		}
	}
```

Reply:
sure this could work